[Bug] Compare fails with Arrow-backed empty join keys: ArrowInvalid: cannot construct ChunkedArray from empty vector and omitted type

datacompy.Compare fails when comparing DataFrames with Arrow-backed join columns in an empty-input edge case.
Internally, Compare calls pandas.merge. In the Arrow factorization path, pandas ends up calling:
pa.chunked_array(lk.chunks + rk.chunks)
If both sides contribute zero chunks, this becomes an empty vector with no explicit type and pyarrow raises:
pyarrow.lib.ArrowInvalid: cannot construct ChunkedArray from empty vector and omitted type

**Versions**

- pyarrow Version: 24.0.0
- datacompy Version: 0.19.5
- pandas Version: 2.3.3

**Observed stack** 
datacompy.core.Compare._dataframe_merge
-> pandas.DataFrame.merge
-> pandas.core.reshape.merge._factorize_keys
-> pa.chunked_array(lk.chunks + rk.chunks)
-> ArrowInvalid

**Expected behavior**
Compare should handle empty Arrow-backed join-key inputs and return a valid comparison result (no exception).

**Actual behavior**
Exception raised during merge/factorization before comparison result is produced.
Minimal reproduction

**Example**
```
# pyarrow Version: 24.0.0
# datacompy Version: 0.19.5
# pandas Version: 2.3.3

# repro_arrowinvalid_minimal.py
import pandas as pd
import pyarrow as pa
import datacompy.core as datacompy

# Control: typed empty Arrow keys -> works
typed = pd.arrays.ArrowExtensionArray(pa.chunked_array([], type=pa.string()))
df1 = pd.DataFrame({"k": pd.Series(typed), "v": pd.Series(typed)})
df2 = pd.DataFrame({"k": pd.Series(typed), "v": pd.Series(typed)})

print("control merge:", df1.merge(df2, on=["k"], how="outer").shape)
print("control datacompy:", datacompy.Compare(df1, df2, join_columns=["k"]).matches())

# Failing primitive used by pandas _factorize_keys in your traceback:
#   pa.chunked_array(lk.chunks + rk.chunks)
# If both sides contribute no chunks and type is omitted -> ArrowInvalid
lk = pa.chunked_array([], type=pa.string())   # 0 chunks
rk = pa.chunked_array([], type=pa.string())   # 0 chunks

print("lk chunks:", lk.num_chunks, "rk chunks:", rk.num_chunks)

# This line reproduces your exact exception message
pa.chunked_array(lk.chunks + rk.chunks)
```

Hi @star11293 , @fdosani,

I checked again with Python 3.11.9 and


```
pandas   2.3.3
datacompy  0.19.5
pyarrow    24.0.0
pyarrow-hotfix  0.7
```


I created a test on our side:

```
    def test_datacompy_arrow_empty_join_keys_regression(self):
        """
        Regression test for pandas/pyarrow/datacompy upgrade compatibility.
        Known failing edge case: two Arrow-backed empty join keys.
        If this test fails with ArrowInvalid, dependency combination is affected.
        """
        arrow_str = pd.ArrowDtype(pa.string())

        df1 = pd.DataFrame(
            {
                "KEY1": pd.Series([], dtype=arrow_str),
                "KEY2": pd.Series([], dtype=arrow_str),
                "VALUE": pd.Series([], dtype=arrow_str),
            }
        )
        df2 = pd.DataFrame(
            {
                "KEY1": pd.Series([], dtype=arrow_str),
                "KEY2": pd.Series([], dtype=arrow_str),
                "VALUE": pd.Series([], dtype=arrow_str),
            }
        )

        # Intentionally no try/except: this test should fail in affected versions
        # and pass once dependency/library behavior is fixed.
        compare = datacompy.Compare(
            df1=df1,
            df2=df2,
            join_columns=["KEY1", "KEY2"],
            ignore_spaces=True,
            ignore_case=False,
        )
        assert compare is not None
```

which fails with the  above stated setup:

```
C:\Users\\.venvs\af_3.2\Scripts\python.exe "C:/Program Files/JetBrains/PyCharm/plugins/python-ce/helpers/pycharm/_jb_pytest_runner.py" --target libs/external/test_datacompy.py::TestDataCompy.test_datacompy_arrow_empty_join_keys_regression 
Testing started at 15:17 ...
Launching pytest with arguments libs/external/test_datacompy.py::TestDataCompy::test_datacompy_arrow_empty_join_keys_regression --no-header --no-summary -q in C:\Users\\Python\DataPipelines\tests

============================= test session starts =============================
collecting ... collected 1 item

libs\external\test_datacompy.py::TestDataCompy::test_datacompy_arrow_empty_join_keys_regression 

============================== 1 failed in 1.19s ==============================
FAILED [100%]
tests\libs\external\test_datacompy.py:133 (TestDataCompy.test_datacompy_arrow_empty_join_keys_regression)
self = <DataPipelines.tests.libs.external.test_datacompy.TestDataCompy object at 0x000001B5BC6E8A90>

    def test_datacompy_arrow_empty_join_keys_regression(self):
        """
        Regression test for pandas/pyarrow/datacompy upgrade compatibility.
        Known failing edge case: two Arrow-backed empty join keys.
        If this test fails with ArrowInvalid, dependency combination is affected.
        """
        arrow_str = pd.ArrowDtype(pa.string())
    
        df1 = pd.DataFrame(
            {
                "KEY1": pd.Series([], dtype=arrow_str),
                "KEY2": pd.Series([], dtype=arrow_str),
                "VALUE": pd.Series([], dtype=arrow_str),
            }
        )
        df2 = pd.DataFrame(
            {
                "KEY1": pd.Series([], dtype=arrow_str),
                "KEY2": pd.Series([], dtype=arrow_str),
                "VALUE": pd.Series([], dtype=arrow_str),
            }
        )
    
        # Intentionally no try/except: this test should fail in affected versions
        # and pass once dependency/library behavior is fixed.
>       compare = datacompy.Compare(
            df1=df1,
            df2=df2,
            join_columns=["KEY1", "KEY2"],
            ignore_spaces=True,
            ignore_case=False,
        )

libs\external\test_datacompy.py:158: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
..\..\..\.venvs\af_3.2\Lib\site-packages\datacompy\core.py:154: in __init__
    self._compare(ignore_spaces=ignore_spaces, ignore_case=ignore_case)
..\..\..\.venvs\af_3.2\Lib\site-packages\datacompy\core.py:250: in _compare
    self._dataframe_merge(ignore_spaces)
..\..\..\.venvs\af_3.2\Lib\site-packages\datacompy\core.py:320: in _dataframe_merge
    outer_join = self.df1.merge(
..\..\..\.venvs\af_3.2\Lib\site-packages\pandas\core\frame.py:10859: in merge
    return merge(
..\..\..\.venvs\af_3.2\Lib\site-packages\pandas\core\reshape\merge.py:184: in merge
    return op.get_result(copy=copy)
           ^^^^^^^^^^^^^^^^^^^^^^^^
..\..\..\.venvs\af_3.2\Lib\site-packages\pandas\core\reshape\merge.py:886: in get_result
    join_index, left_indexer, right_indexer = self._get_join_info()
                                              ^^^^^^^^^^^^^^^^^^^^^
..\..\..\.venvs\af_3.2\Lib\site-packages\pandas\core\reshape\merge.py:1152: in _get_join_info
    (left_indexer, right_indexer) = self._get_join_indexers()
                                    ^^^^^^^^^^^^^^^^^^^^^^^^^
..\..\..\.venvs\af_3.2\Lib\site-packages\pandas\core\reshape\merge.py:1126: in _get_join_indexers
    return get_join_indexers(
..\..\..\.venvs\af_3.2\Lib\site-packages\pandas\core\reshape\merge.py:1741: in get_join_indexers
    zipped = zip(*mapped)
             ^^^^^^^^^^^^
..\..\..\.venvs\af_3.2\Lib\site-packages\pandas\core\reshape\merge.py:1738: in <genexpr>
    _factorize_keys(left_keys[n], right_keys[n], sort=sort)
..\..\..\.venvs\af_3.2\Lib\site-packages\pandas\core\reshape\merge.py:2486: in _factorize_keys
    pa.chunked_array(lk.chunks + rk.chunks)  # type: ignore[union-attr]
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pyarrow/table.pxi:1555: in pyarrow.lib.chunked_array
    ???
pyarrow/error.pxi:155: in pyarrow.lib.pyarrow_internal_check_status
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   ???
E   pyarrow.lib.ArrowInvalid: cannot construct ChunkedArray from empty vector and omitted type

pyarrow/error.pxi:92: ArrowInvalid

```

--- a/datacompy/pandas.py
+++ b/datacompy/pandas.py
@@ -363,11 +363,16 @@ class PandasCompare(BaseCompare):
         else:
             params = {"on": self.join_columns}
 
-        for column in self.join_columns:
-            self.df1[column] = pandas_normalize_string_column(
-                self.df1[column], ignore_spaces=ignore_spaces, ignore_case=False
-            )
-            self.df2[column] = pandas_normalize_string_column(
+        # Skip normalization for empty frames: str.strip() on an empty
+        # Arrow-backed column drops its chunks to zero, which causes pandas
+        # merge to call pa.chunked_array([]) and raise ArrowInvalid when
+        # multiple join columns are present. See issue #514.
+        if len(self.df1) > 0 or len(self.df2) > 0:
+            for column in self.join_columns:
+                self.df1[column] = pandas_normalize_string_column(
+                    self.df1[column], ignore_spaces=ignore_spaces, ignore_case=False
+                )
+                self.df2[column] = pandas_normalize_string_column(
                 self.df2[column], ignore_spaces=ignore_spaces, ignore_case=False
             )
 

--- a/datacompy/pandas.py
+++ b/datacompy/pandas.py
@@ -373,8 +373,8 @@ class PandasCompare(BaseCompare):
                     self.df1[column], ignore_spaces=ignore_spaces, ignore_case=False
                 )
                 self.df2[column] = pandas_normalize_string_column(
-                self.df2[column], ignore_spaces=ignore_spaces, ignore_case=False
-            )
+                    self.df2[column], ignore_spaces=ignore_spaces, ignore_case=False
+                )
 
         outer_join = self.df1.merge(
             self.df2,

--- a/tests/test_pandas.py
+++ b/tests/test_pandas.py
@@ -2645,6 +2645,40 @@ def test_columns_with_mismatches_empty_dataframes():
     assert result == []
 
 
+def test_compare_empty_arrow_backed_join_keys():
+    """Regression for #514: empty Arrow-backed join keys must not crash merge."""
+    import pyarrow as pa
+
+    for arrow_type in (pa.string(), pa.int64()):
+        dtype = pd.ArrowDtype(arrow_type)
+        df1 = pd.DataFrame(
+            {
+                "key1": pd.Series([], dtype=dtype),
+                "key2": pd.Series([], dtype=dtype),
+                "value": pd.Series([], dtype=dtype),
+            }
+        )
+        df2 = pd.DataFrame(
+            {
+                "key1": pd.Series([], dtype=dtype),
+                "key2": pd.Series([], dtype=dtype),
+                "value": pd.Series([], dtype=dtype),
+            }
+        )
+
+        compare = PandasCompare(
+            df1, 
+            df2, 
+            join_columns=["key1", "key2"], 
+            ignore_spaces=True,
+            ignore_case=False
+        )
+
+        assert len(compare.intersect_rows) == 0
+        assert len(compare.df1_unq_rows) == 0
+        assert len(compare.df2_unq_rows) == 0
+
+
 def test_columns_with_mismatches_on_index():
     """Test columns_with_mismatches when comparing on index."""
     df1 = pd.DataFrame(

--- a/tests/test_pandas.py
+++ b/tests/test_pandas.py
@@ -2667,11 +2667,11 @@ def test_compare_empty_arrow_backed_join_keys():
         )
 
         compare = PandasCompare(
-            df1, 
-            df2, 
-            join_columns=["key1", "key2"], 
+            df1,
+            df2,
+            join_columns=["key1", "key2"],
             ignore_spaces=True,
-            ignore_case=False
+            ignore_case=False,
         )
 
         assert len(compare.intersect_rows) == 0

